File: 300. Longest Increasing Subsequence - OPT.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        ArrayList<Integer> arr = new ArrayList<>();
        arr.add(nums[0]);
        for(int i = 1; i < nums.length; i++){
            if(arr.get(arr.size() - 1) < nums[i]) arr.add(nums[i]);
            else{
                int left = 0;
                int right = arr.size();
                while(left < right){
                    int mid = left + (right - left) / 2;
                    if(arr.get(mid) >= nums[i]) right = mid;
                    else left = mid + 1;
                }
                if(left < arr.size()) arr.set(left, nums[i]);
            }
        }

        return arr.size();
    }
}

File: 300. Longest Increasing Subsequence -OPT.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> arr;
        for (int num : nums) {
            if (arr.empty() || arr.back() < num) {
                arr.push_back(num);
            } else {
                int left = 0, right = (int)arr.size() - 1;
                int ans = 0;
                while (left <= right) {
                    int mid = left + (right - left) / 2;
                    if (arr[mid] < num) {
                        left = mid + 1;
                    } else {
                        ans = mid;
                        right = mid - 1;
                    }
                }
                arr[ans] = num;
            }
        }
        return (int)arr.size();
    }
};